### ISSUE TITLE
(aws-ec2): Add custom allocationId support to Gateway NatProvider

### ISSUE THREAD

**Issue by TikiTDO** [aws/aws-cdk#11884]
The default `NatProvider.gateway()` generates new EIP allocations when creating the gateway. This feature will allow you use an existing EIP allocation which may not be replaceable due to business reasons.

### Use Case

Client has a set of EIPs whitelisted with multiple vendors. I want to be able to use the existing EIPs while moving to CDK for most of the infrastructure. 

### Proposed Solution

Add an optional `allocationIds` parameter to the `NatGatewayProvider` class and `NatProvider.gateway()` helper function. PR available at #11885

As a more configurable alternative, could modify `PublicSubnetProps` and `PublicSubnet` to add a `natGatewayAllocationId` in order to allow configuring specific EIP allocations for specific subnets.

---

This is a :rocket: Feature Request

**Comment by rix0rrr** [comment]
The CDK is not going to contain and maintain every possible combination of features in its core library. That's why it's designed to be extensible.

Please use your own custom subclass of `NatProvider` for your customer's project.

If you want to share it with other people, put it into your own construct library and publish that for yourself. If there's enough traction and proven demand for your custom `NatProviderWithCustomAllocations`, we will reconsider adding it to the core library.

**Comment by TikiTDO** [comment]
@rix0rrr Is there a set of guidelines as to what is considered an "acceptable" feature, and how much proven demand is necessary to approve a feature? I would certainly contend that there is at least some measurable demand already.

From my own experience, I have had to assign existing EIPs to NAT gateways for several clients in totally different industries already, so in my sample space this is a feature that would get quite a bit of use. IP based whitelisting is rather common in multiple industries (medicine, finance, insurance), and any client in such industries with existing AWS resources is likely to have existing allocations which they can not lose without having to engage multiple vendors.

In fact the ability to select an EIP is one of the 3 fields displayed on the "Create NAT gateway" page, with the ability to allocate a new EIP being a secondary button, so it appears that the team in charge of the VPC dashboard considers it to be an important feature of the service.

![image](https://user-images.githubusercontent.com/326184/101514926-cbd50d80-394b-11eb-9704-dd5f842f2ad9.png)

In terms of popularity; it seems that you have already closed a few issues that have asked for this very same feature, which suggests to me that there is at least some proven demand, given that at least 4 different people have now spent the time and effort necessary to create these issues, and even more have participated. At least one of these issues has sparked a decent amount of discussion, with multiple people asking questions and offering examples over the course of a year. Such continuing discussion is usually indicative of a consistent pain point.

#4067
#4705
#8327

I'm also not sure that publishing a custom library can prove out demand for a feature like this any more so than looking at the issue board in this repo. 

Looking online, there seems to be essentially no market for standalone custom CDK constructs, with little to no community, practically no discoverability, no security guidelines for authors, and almost nothing to steer people with problems towards custom third-party constructs. As a result, while custom constructs are a great tool to DRY up infrastructure stacks and formalize design paradigms within an organization, I have yet to see even a single usage of an external non-CDK construct by even a single one of my clients. In fact you are probably the first person I've actually seen to suggest releasing a standalone CDK construct as an NPM module. 

Frameworks such as [punchcard](https://github.com/punchcard/punchcard) come the closest with an extra layer of high-level constructs, though these constructs also seem to be more about formalizing design patters in the framework. 

In other words, since CDK already offers most of the primitive functionality necessary to generate almost any valid CloudFormation template, there is simply very little market to consume third party code, particularly when it comes to something as sensitive as infrastructure.

Realistically, anyone that hits this issue will just spent a few days cursing until they either find an older issue with these constructs, or figure out how to make a custom construct that meets their own needs. Having CDK provide this functionality would remove this barrier, while aligning closer with the user experience offered by the AWS dashboard.
